src/writer.cc

```diff
@@ -22,177 +22,123 @@
-  memcpy(writer->buf.start + offset, data, size);
-  if (end > writer->buf.size)
-    writer->buf.size = end;
+Result MemoryWriter::WriteData(size_t dst_offset,
+                               const void* src,
+                               size_t size) {
+  size_t end = dst_offset + size;
+  if (end > buf_->data.size()) {
+    buf_->data.resize(end);
+  }
+  uint8_t* dst = &buf_->data[dst_offset];
+  memcpy(dst, src, size);
   return Result::Ok;
 }
 
-static Result move_data_in_output_buffer(size_t dst_offset,
-                                         size_t src_offset,
-                                         size_t size,
-                                         void* user_data) {
-  MemoryWriter* writer = static_cast<MemoryWriter*>(user_data);
+Result MemoryWriter::MoveData(size_t dst_offset,
+                              size_t src_offset,
+                              size_t size) {
+  if (size == 0) {
+    return Result::Ok;
+  }
+
   size_t src_end = src_offset + size;
   size_t dst_end = dst_offset + size;
   size_t end = src_end > dst_end ? src_end : dst_end;
-  ensure_output_buffer_capacity(&writer->buf, end);
-  void* dst = reinterpret_cast<void*>(
-      reinterpret_cast<size_t>(writer->buf.start) + dst_offset);
-  void* src = reinterpret_cast<void*>(
-      reinterpret_cast<size_t>(writer->buf.start) + src_offset);
+  if (end > buf_->data.size()) {
+    buf_->data.resize(end);
+  }
+
+  uint8_t* dst = &buf_->data[dst_offset];
+  uint8_t* src = &buf_->data[src_offset];
   memmove(dst, src, size);
-  if (end > writer->buf.size)
-    writer->buf.size = end;
   return Result::Ok;
 }
 
-Result init_mem_writer(MemoryWriter* writer) {
-  WABT_ZERO_MEMORY(*writer);
-  writer->base.user_data = writer;
-  writer->base.write_data = write_data_to_output_buffer;
-  writer->base.move_data = move_data_in_output_buffer;
-  init_output_buffer(&writer->buf, INITIAL_OUTPUT_BUFFER_CAPACITY);
-  return Result::Ok;
-}
+FileWriter::FileWriter(FILE* file)
+    : file_(file), offset_(0), should_close_(false) {}
 
-Result init_mem_writer_existing(MemoryWriter* writer, OutputBuffer* buf) {
-  WABT_ZERO_MEMORY(*writer);
-  writer->base.user_data = writer;
-  writer->base.write_data = write_data_to_output_buffer;
-  writer->base.move_data = move_data_in_output_buffer;
-  writer->buf = *buf;
-  /* Clear buffer, since ownership has passed to the writer. */
-  WABT_ZERO_MEMORY(*buf);
-  return Result::Ok;
-}
+FileWriter::FileWriter(const char* filename) : file_(nullptr), offset_(0) {
+  file_ = fopen(filename, "wb");
 
-void steal_mem_writer_output_buffer(MemoryWriter* writer,
-                                    OutputBuffer* out_buf) {
-  *out_buf = writer->buf;
-  writer->buf.start = nullptr;
-  writer->buf.size = 0;
-  writer->buf.capacity = 0;
+  // TODO(binji): this is pretty cheesy, should come up with a better API.
+  if (!file_) {
+    ERROR("fopen name=\"%s\" failed, errno=%d\n", filename, errno);
+  }
+  should_close_ = file_ != nullptr;
```

Comment:
should_close_(false) in the init list and then set to true here (or in else block if you prefer)?   since you already check for !file_ above.